java/code/src/com/suse/manager/webui/controllers/contentmanagement/handlers/ProjectSourcesApiController.java

```diff
@@ -0,0 +1,110 @@
+
+import static com.suse.manager.webui.utils.SparkApplicationHelper.json;
+import static com.suse.manager.webui.utils.SparkApplicationHelper.withUser;
+import static spark.Spark.put;
+
+import com.redhat.rhn.domain.contentmgmt.ContentEnvironment;
+import com.redhat.rhn.domain.contentmgmt.ContentProject;
+import com.redhat.rhn.domain.contentmgmt.ProjectSource;
+import com.redhat.rhn.domain.contentmgmt.SoftwareProjectSource;
+import com.redhat.rhn.domain.user.User;
+import com.redhat.rhn.manager.contentmgmt.ContentManager;
+
+import com.suse.manager.webui.controllers.contentmanagement.mappers.ResponseMappers;
+import com.suse.manager.webui.controllers.contentmanagement.request.ProjectSourcesRequest;
+import com.suse.manager.webui.utils.gson.ResultJson;
+import com.suse.utils.Json;
+
+import com.google.gson.Gson;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import spark.Request;
+import spark.Response;
+
+/**
+ * Spark controller class for content management sources API endpoints.
+ */
+public class ProjectSourcesApiController {
+
+    private static final Gson GSON = Json.GSON;
+
+    private ProjectSourcesApiController() {
+    }
+
+    /** Init routes for ContentManagement Sources Api.*/
+    public static void initRoutes() {
+        put("/manager/contentmanagement/api/projects/:projectId/softwaresources",
+                withUser(ProjectSourcesApiController::updateContentSoftwareSources));
+
+    }
+
+    /**
+     * Return the JSON with the result of updating the content project with a list of sources.
+     * @param req the http request
+     * @param res the http response
+     * @param user the current user
+     * @return the JSON data
+     */
+    public static String updateContentSoftwareSources(Request req, Response res, User user) {
+        ProjectSourcesRequest createSourceRequest = ProjectSourcesHandler.getSourcesRequest(req);
+        String projectLabel = createSourceRequest.getProjectLabel();
+
+        ContentProject dbContentProject = ContentManager.lookupProject(
+                createSourceRequest.getProjectLabel(), user
+        ).get();
+
+        List<String> sourceLabelsToDetach = dbContentProject.getSources()
+                .stream()
+                .filter(SoftwareProjectSource.class::isInstance)
+                .map(SoftwareProjectSource.class::cast)
+                .map(softwareSource -> softwareSource.getChannel().getLabel())
+                .collect(Collectors.toList());
+        sourceLabelsToDetach.forEach(sourceLabel -> ContentManager.detachSource(
+                projectLabel,
+                ProjectSource.Type.SW_CHANNEL,
+                sourceLabel,
+                user
+        ));
+
+        List<String> sourceLabelsToAttach = createSourceRequest.getSoftwareSources()
+                .stream()
+                .map(source -> source.getLabel())
+                .collect(Collectors.toList());
+        int sourcePosition = 0;
```

Comment:
Is the `sourcePosition` there only because you want to add the sources one-after-one in the order that you receive them? In this case, it'd be maybe nicer to reverse the collection and call `attachSource` with `position == empty()`.

Reply:
humm I don't think this will work if we have this code: 

https://github.com/uyuni-project/uyuni/pull/686/commits/fb0a1c1d31754f7152157ccdefa3758dfa345de9#diff-8b190e9f8affc17d84c97efbc55fb8e6R260

Should we assume that if the position isn't present we just remove and add the source to the first position?

Reply:
I think now it's good :)